### ISSUE TITLE
QueryBuilder does not appropriately route Postgres connections with differing adapter class names

### ISSUE THREAD

**Issue by jcmcken** [wvanbergen/scoped_search#139]
I'm attempting to use the ``makara`` rubygem to shard database reads and writes to different database servers. This gem provides its own alternate connection adapter that wraps whatever default adapter you're using to intercept and route the request based on what SQL is being executed. Because``ScopedSearch::QueryBuilder`` builds SQL differently depending on the name of the connection adapter class (https://github.com/wvanbergen/scoped_search/blob/master/lib/scoped_search/query_builder.rb#L30), and because my connection adapter is not known (``MakaraPostgreSQLAdapter`` rather than ``PostgreSQLAdapter``), ``scoped_search`` generates invalid Postgres SQL which causes the app to raise a PGError.

The specifics of the error aren't important in this case, but just to illustrate, ``scoped_search`` is giving me the SQL...

```sql
SELECT COUNT(*) FROM "hosts"  WHERE "hosts"."type" IN ('Host::Managed') AND (((NOT COALESCE("hosts"."last_report" IS NOT NULL, 0)) AND (("hosts"."enabled" <> 'f'))))
```

...rather than:

```sql
SELECT COUNT(*) FROM "hosts"  WHERE "hosts"."type" IN ('Host::Managed') AND (((NOT COALESCE("hosts"."last_report" IS NOT NULL, false)) AND (("hosts"."enabled" <> 'f'))))
```

In Postgres, the first statement raises a PGError at the ``COALESCE``, because it detects that that part of the statement should resolve to a boolean, but is instead resolving to an integer (``0``)

**Comment by domcleal** [comment]
https://groups.google.com/d/msg/foreman-users/-7JdhZ9XT7k/I_iuqUZsCQAJ for the record.

Adding `class MakaraPostgreSQLAdapter < PostgreSQLAdapter; end` is probably sufficient.

**Comment by jcmcken** [comment]
That works. Rather than support specific adapter names, would it make sense to have a heuristic for determining if the adapter is a PG adapter instead? Like:

```ruby
def self.class_for(definition)
  case definition.klass.connection.class.name.split('::').last
  when /postgresql/i
    PostgreSQLAdapter
  else
    self
  end
end
```

Not foolproof, but it would catch more cases at the least.

**Comment by domcleal** [comment]
Yeah, that'd probably work OK - though I'm not too aware of what other adapters are out in the wild.

**Comment by wvanbergen** [comment]
fixed by #140
